csharp/src/Google.Protobuf/ParsingPrimitives.cs

```diff
@@ -453,36 +448,13 @@ public static byte[] ReadRawBytes(ref ReadOnlySpan<byte> buffer, ref ParserInter
                 state.bufferPos += size;
                 return bytes;
             }
-            else if (size < buffer.Length || size < state.segmentedBufferHelper.TotalLength)
+            else if (IsDataAvailableInSource(ref state, size))
```

Comment:
It makes sense to change `size < state.segmentedBufferHelper.TotalLength` part of the condition into the more strict `IsDataAvailableInSource(ref state, size)` (which also takes into account the amount of data we've read) because we should never be reading anything over the known TotalLength.

the `size < buffer.Length` part of the condition is missing now  and I think that was there to express that if we have CodedInputStream that reads from stream, the buffer being used is considered small enough so that allocating a new byte array up to its size is still safe.

Because we want to avoid taking the code path of allocating chunk-by-chunk and then concatenating those,
it think we can make this condition:
`(!state.segmentedBufferHelper.TotalLength.HasValue &&  size < buffer.Length) || IsDataAvailableInSource(ref state, size)`
that ways we will still be conservative when totalLength is known but we will allow allocating a reasonable sized array when reading from Stream with an intermediate buffer. (this also preserves the original behavior in that case).